File: GrokkingDP/5/MaxSumIncSubseq.cpp
#include <iostream>
#include <vector>

enum class Method { BRUTE, MEM, BU };

int MaxSumIncSubseq(std::vector<int>& nums, Method m);
int Brute(std::vector<int>& nums, int i, int j);
int Memoize(std::vector<int>& nums, int i, int j,
            std::vector<std::vector<int>>& dp);
int BU(std::vector<int>& nums);

int main(void) {
  std::vector<int> s1 = {4, 1, 2, 6, 10, 1, 12};
  std::vector<int> s2 = {-4, 10, 3, 7, 15};
  Method m = Method::BU;
  std::cout << MaxSumIncSubseq(s1, m) << ":" << MaxSumIncSubseq(s2, m);
  return 0;
}

int MaxSumIncSubseq(std::vector<int>& nums, Method m) {
  std::vector<std::vector<int>> dp(nums.size(), std::vector<int>(nums.size()));
  switch (m) {
    case Method::BRUTE:
      return Brute(nums, 0, -1);
    case Method::MEM:
      return Memoize(nums, 0, -1, dp);
    case Method::BU:
      return BU(nums);
  }
}

int Brute(std::vector<int>& nums, int i, int j) {
  if (i == (int)nums.size()) return 0;
  int include = 0;
  if (j == -1 || nums[i] > nums[j]) include = nums[i] + Brute(nums, i + 1, i);
  return std::max(include, Brute(nums, i + 1, j));
}

int Memoize(std::vector<int>& nums, int i, int j,
            std::vector<std::vector<int>>& dp) {
  if (i == (int)nums.size()) return 0;
  if (dp[i][j + 1] == 0) {
    int include = 0;
    if (j == -1 || nums[i] > nums[j])
      include = nums[i] + Memoize(nums, i + 1, i, dp);
    dp[i][j + 1] = std::max(include, Memoize(nums, i + 1, j, dp));
  }
  return dp[i][j + 1];
}

int BU(std::vector<int>& nums) {
  std::vector<int> dp(nums.size());
  int n = nums.size();
  int largest = 0;
  for (int i = 0; i < n; ++i) {
    dp[i] = nums[i];
    largest = std::max(largest, nums[i]);
  }

  for (int j = 1; j < n; ++j) {
    for (int k = 0; k < j; ++k) {
      int sum = dp[k] + nums[j];
      if (nums[j] > nums[k] && sum > dp[j]) dp[j] = sum;
    }
    largest = std::max(largest, dp[j]);
  }

  return largest;
}

File: GrokkingDP/5/max_sum_inc_subseq.py
#  Problem Statement #
#  Given a number sequence, find the increasing subsequence with the highest sum. Write a method that returns the highest sum.

#  Example 1:

#  Input: {4,1,2,6,10,1,12}
#  Output: 32
#  Explanation: The increaseing sequence is {4,6,10,12}.
#  Please note the difference, as the LIS is {1,2,6,10,12} which has a sum of '31'.

#  Example 2:

#  Input: {-4,10,3,7,15}
#  Output: 25
#  Explanation: The increaseing sequences are {10, 15} and {3,7,15}.

from typing import List

def td(nums: List[int]) -> int:
    """
    >>> td([4,1,2,6,10,1,12])
    32
    >>> td([-4,10,3,7,15])
    25
    """
    N = len(nums)
    dp = [[0 for _ in range(N+1)] for _ in range(N)]
    def fn(i: int, j: int) -> int:
        if i == N: return 0
        if j == -1 or nums[i] > nums[j]: dp[i][j+1] = nums[i] + fn(i+1, i)
        dp[i][j+1] = max(fn(i+1, j), dp[i][j+1])
        return dp[i][j+1]
    return fn(0, -1)

def bu(nums: List[int]) -> int:
    """
    >>> bu([4,1,2,6,10,1,12])
    32
    >>> bu([-4,10,3,7,15])
    25
    """
    N, mx = len(nums), 0
    dp = nums[:]
    for i in range(1, N):
        for j in range(i):
            if nums[i] > nums[j]: dp[i] = max(dp[i], nums[i] + dp[j])
        mx = max(mx, dp[i])
    return mx

if __name__ == '__main__':
    import doctest
    doctest.testmod()
